### PR TITLE
[Feature] alphabet name

### PR DESCRIPTION
**Pull request by rrahn**
Adds a name field to the alphabet concept and all implemented alphabets.
It uses the `static_string` from #131 to assemble names at compile time.
This will be used in the tokenisation to give the user pretty names to better understand what went wrong.
Like `The value 'N' is not part of the alphabet 'dna4'`.

Regarding the static_string please look into #131  first.
This PR is rebased on #131 .

###ToDO
 - [ ] docu

### CODE REVIEW

**Comment by h-2** [comment]
I like the static string class, but I am reluctant about the general proposal. This functionality emulates introspection in an over-verbose manner, from my POV. I agree the feature is nice to have in IO, but it adds no functionality to the alphabet itself and it is unclear, why e.g. alphabets do this, but other types don't. I would propose two alternatives:
  * Use actual (weak) introspection a la [`typeid(l).name()`](http://en.cppreference.com/w/cpp/types/type_info/name) ; *or if used more often:*
  * Define `seqan3::typename_as_string()` that does introspection and/or can be overloaded to provide a user-defined string

I don't know if this is relevant to your current work, but please bear in mind, that while all current alphabets know their letters at compile-time, this doesn't necessarily have to to be true and in fact I have some alphabets up my sleeve where it isn't :smile:
[the size on the other hand is required to be `constexpr`]

**Comment by h-2** [comment]
Let's discuss this on Monday!

**Comment by h-2** [comment]
A dedicated function could also employ demangling if available...

**Comment by rrahn** [comment]
Respectfully, I disagree, but let's discuss it on Monday sure.

**Comment by rrahn** [comment]
Still not completely convinced, but I agree it would be better to wait for reflection.
So instead of doing things `constexpr`, I do it the not so cool way. 😞

**Comment by h-2** [comment]
> So instead of doing things constexpr, I do it the not so cool way. :disappointed:

I have done some more reading on this:
* http://www.bfilipek.com/2018/02/staticvars.html
* https://meowni.ca/posts/static-initializers/
* https://isocpp.org/wiki/faq/ctors#static-init-order
* http://neugierig.org/software/chromium/notes/2011/08/static-initializers.html

TL;DR: `static const std::string =` is discouraged in general (or used to be), because it causes problems with multiple translation units. I do think that this is mitigated with inline variables, because cppreference explicitly states:
> If the initialization of an inline variable is deferred, it happens before the first odr-use of that specific variable.  (since C++17)

And also all the variables being in templates and having internal linkage should just have them appear in each translation unit separately IIANM.

But I would feel better if you also have a look at these resources.

----

Alternatively there seems to be a cool hack to get a `constexpr` version of typeid:
https://stackoverflow.com/questions/35941045/can-i-obtain-c-type-names-in-a-constexpr-way/35943472#35943472
Using this you could keep the constexpr string stuff, but relying on `__PRETTY_FUNCTION__` is even more craziness than typeid.

**Comment by h-2** [comment]
Another post that explains why `inline` variables actually do fix this:
https://stackoverflow.com/a/38043566

So it looks like it's ok to do it like we discussed and you implemented in #137 .

**Comment by rrahn** [comment]
Well, it really is still bad I think. 
Maybe I might misunderstood the inline variable. But from what I could read it seems it is quite possible that multiple translation units could potentially evaluate the `typeid` name differently and thus would invalidate the invariant that every TU defines the inline variable with the exact same value.
Does it not?
You got me hooked with `__PRETTY_FUNCTION__` though.

**Comment by rrahn** [comment]
I mean I am not saying it is likely or anything but, frankly, it is just not guaranteed that this will not happen.

**Comment by marehr** [comment]
We could still use the `static reflection` proposal

```cpp
std::cout << "type name:" << get_name_v<type> << "\n";
```
:evergreen_tree: 

http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0194r4.html

**Comment by rrahn** [comment]
The proposal would be the right way to go, but as @h-2 mentioned, the current version would require that every alphabet needs to be touched and that the concept must be adapted. But once reflection is in, we don't need it as a concept, breaking the compatibility, which I would like to avoid. Then again we could use a static string together with the `__PRETTY_FUNCTION__` of gcc. I think that is quite reasonable to do and resolves both our major concerns. A) evaluate it at compile time and B) don't do it as a name filed in the class itself, but do it generically.

**Comment by h-2** [comment]
@rrahn I thought you already implemented it :laughing:
